File: 744-network-delay-time/network-delay-time.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<vector<int>> grid(n+1, vector<int>(n+1, INT_MAX));
        for (int i=0; i<times.size(); i++) {
            int s = times[i][0], t = times[i][1], w = times[i][2];
            grid[s][t] = w;
        }

        vector<bool> visited(n+1, false);
        vector<int> dist(n+1, INT_MAX);

        // start from k
        dist[k] = 0;

        // 1. each time select 1 node
        for (int i=1; i<=n; i++) {  // traverse each node
            int minDist = INT_MAX;
            int cur = k;
            for (int j=1; j<=n; j++) {
                // select non-visited node
                if (!visited[j] && dist[j] < minDist) {
                    minDist = dist[j];
                    cur = j;
                }
            }
            // 2. mark as visited
            visited[cur] = true;

            // 3. update the dist array
            for (int w=1; w<=n; w++) {
                if (!visited[w] && grid[cur][w] != INT_MAX && dist[cur] + grid[cur][w] < dist[w]) {
                    dist[w] = dist[cur] + grid[cur][w];
                }
            }
        }

        // real answer: finding maximum
        int ans = -1;
        for (int i=1; i<=n; i++) {
            ans = max(ans, dist[i]);
        }
        return (ans == INT_MAX) ? -1 : ans;
    }
};

File: 744-network-delay-time/network-delay-time.py
class Solution(object):
    def networkDelayTime(self, times, n, k):
        """
        :type times: List[List[int]]
        :type n: int
        :type k: int
        :rtype: int
        """
        # build adjacency list
        graph = [[] for _ in range(n+1)]
        for s, t, v in times:
            graph[s].append((t, v))

        # create distance array
        INT_MAX = 1e9
        minDist = [INT_MAX] * (n+1)

        # optimized: if the element is in the queue, don't append twice
        InQueue = [False] * (n+1)

        # queue optimized bellmen ford
        q = deque()
        q.append(k)
        minDist[k] = 0
        InQueue[k] = True

        while q:    # !q.empty()
            src = q.popleft()
            InQueue[src] = False

            # relax all outgoing edges
            for t, v in graph[src]:
                # update the neighbor
                if minDist[src] + v < minDist[t]:
                    minDist[t] = minDist[src] + v
                    if not InQueue[t]:
                        InQueue[t] = True
                        q.append(t)

        ans = 0
        for i in range(1, n+1):
           
            if minDist[i] == INT_MAX:
                return -1

            ans = max(ans, minDist[i])

        return ans
